### DISCUSSION TITLE
Procedure to convert from ASCII to Binary ?

### DISCUSSION THREAD

**Discussion by smoluck** [dranjan/python-plyfile#50] [Q&A] [answered]
Hi Dranjan.

I've used successfully your module to convert from PYL Binary (little_endian) to ASCII and cleanup some data in the process, but now i want to convert that file sample to Binary (little_endian) again.

How can i do that ?

`ply
format ascii 1.0
obj_info Photoneo
element vertex 5
property float x
property float y
property float z
property uchar red
property uchar green
property uchar blue
end_header
224.491729736328125 -707.4791259765625 3324.802490234375 215 215 215 
214.3670654296875 -706.00433349609375 3325.119140625 201 201 201 
216.327590942382812 -707.69073486328125 3333.017822265625 231 231 231 
217.485824584960938 -706.7474365234375 3328.53125 255 255 255 
218.767044067382812 -706.21368408203125 3325.97314453125 249 249 249
 `
 

When i try to read the data of that file i always end up with this error that refer to plyfile.py. Does it have a relation with the file encoding that doesn't need to be decoded from Python 3.7 env ?
 
` line = raw_line.decode('ascii').strip()
AttributeError: 'str' object has no attribute 'decode'
`

**Comment by dranjan** [comment]
Hi @smoluck. Any chance you could post your code, and attach the example file itself? I'd like to try to reproduce the behavior. It's not clear from what you wrote if that error is a `plyfile` bug or something else.

Converting from ASCII to binary is exactly like the opposite conversion, you just set the `.text` and `.byte_order` on the `PlyData` instance before writing it to file.

**Reply by smoluck** [reply]
Here is the code i was trying to use.

I've created the dat_ASCII.ply file by doing a conversion from Binary to ASCII first with some cleanup on the data values.

**Comment by smoluck** [comment]
```
import plyfile

# Code for converting ASCII back to Binary (little_endian)
inputfile = r"C:\TEMP\T\dat_ASCII.ply"
print(inputfile)
outputfile = r"C:\TEMP\T\dat_Binary.ply"
print(outputfile)

file1 = open(inputfile, 'r')
file2 = open(outputfile, 'w+')

plydata = plyfile.PlyData.read(file1)
plydata.text = True
plydata.byte_order = "<"
plydata.write(file2)
# close and save the files
file2.close()
file1.close()
```

**Comment by smoluck** [comment]
the error returned by PyCharm

```
Traceback (most recent call last):
  File "C:\Python\Python37\lib\code.py", line 90, in runcode
    exec(code, self.locals)
  File "<input>", line 26, in <module>
  File "C:\Project\Vir_Env_Python\lib\site-packages\plyfile.py", line 395, in read
    data = PlyData._parse_header(stream)
  File "C:\Project\Vir_Env_Python\lib\site-packages\plyfile.py", line 373, in _parse_header
    while parser.consume(stream.readline()):
  File "C:\Project\Vir_Env_Python\lib\site-packages\plyfile.py", line 120, in consume
    line = raw_line.decode('ascii').strip()
AttributeError: 'str' object has no attribute 'decode'
```

**Comment by smoluck** [comment, marked as answer]
_Hi Dranjan.

Thanks for your feedback.
Best regards,_

I've found my errors.

As i've modified the amount of lines for the data values, i was getting an error too.

```
        ### Important ###
        # As an empty line will always been added in the RAW ASCII file,
        # it's important to remove one to correctly count the Point Count.
        # "element vertex" count will be used by the Binary Encoding at the end of this script
```

Here is the update code that is now working.
**rb** and **wb** flags have also helped here.

```
import os, plyfile

# Code for converting ASCII back to Binary (little_endian)
inputfile = r"C:\TEMP\T\dat_ASCII.ply"
inputfile = os.path.abspath(inputfile)
print(inputfile)

outputfile = r"C:\TEMP\T\dat_Binary.ply"
outputfile = os.path.abspath(outputfile)
print(outputfile)

file1 = open(inputfile, 'rb')
file2 = open(outputfile, 'wb+')

plydata = plyfile.PlyData.read(file1)
plydata.text = False
plydata.byte_order = "<"
plydata.write(file2)
# close and save the files
file2.close()
file1.close()
```

**Comment by dranjan** [comment]
Ah, right. The input file does need to be opened in binary mode. That might need to be emphasized more in the documentation.
